### PR TITLE
SNOW-1562363: Support for `DataFrame/Series.asfreq`

### PR DESCRIPTION
**Pull request by sfc-gh-nkrishna**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   SNOW-1562363

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This PR adds support for `DataFrame/Series.asfreq` by calling the `resample` implementation.

### CODE REVIEW

File: src/snowflake/snowpark/modin/pandas/base.py [hunk 1]
```diff
@@ -1045,21 +1045,25 @@ def error_raiser(msg, exception):
         )
         return query_compiler
 
-    @base_not_implemented()
     def asfreq(
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/pandas/base.py:1048]: can you also update the docstring for this API?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -10519,12 +10518,89 @@ def rank(
...
+            "convention": "start",
+            "kind": None,
+            "on": None,
+            "level": None,
+            "origin": "start_day",
+            "offset": None,
+            "group_keys": no_default,
+        }  # pragma: no cover
+
+        return self.resample(  # pragma: no cover
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10590]: If this line is covered then we don't need "# pragma: no cover"

File: tests/integ/modin/resample/test_resample.py [hunk 3]
```diff
@@ -477,3 +477,30 @@ def test_resample_ffill_large_gaps(interval):
         lambda df: df.resample(rule=f"{interval}D").ffill(),
         check_freq=False,
     )
+
+
+@freq
+@interval
+@sql_count_checker(query_count=3, join_count=1)
```
Comment (sfc-gh-azhan) [inline tests/integ/modin/resample/test_resample.py:484]: why 3 queries? 2 from resample?

Comment (sfc-gh-nkrishna) [reply tests/integ/modin/resample/test_resample.py:484]: All the tests in this file have 3 queries, 1 to convert index to native pandas for dataframe constructor, then 2 from the resample. Resample executes an eager query to get the min and max values of the index.

Below outside of `eval_snowpark_pandas_result` is only 2 queries:
```
    with SqlCounter(query_count=2):
        import modin.pandas as pd
        df = pd.DataFrame({"A": np.random.randn(15)}, index=native_pd.date_range("2020-01-01", periods=15, freq="1s"))
        df.asfreq(freq="5s", method="ffill").to_pandas()
```

File: src/snowflake/snowpark/modin/pandas/base.py (hunk 1 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/pandas/base.py:1048]: Ah forgot, will do.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py]: This was complaining earlier, let me try again.

File: src/snowflake/snowpark/modin/plugin/docstrings/base.py [hunk 4]
```diff
@@ -462,6 +462,50 @@ def apply():
...
+        normalize : bool, default False
+            Whether to reset output index to midnight.
+
+        fill_value : scalar, optional
+            Value to use for missing values, applied during upsampling
+            (note this does not fill NaNs that already were present).
+
+        Returns
+        -------
+        SnowflakeQueryCompiler
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/docstrings/base.py:501]: should be either dataframe or series.

File: docs/source/modin/supported/series_supported.rst [hunk 5]
```diff
@@ -100,7 +100,8 @@ Methods
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``argsort``                 | N                               |                                  |                                                    |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
-| ``asfreq``                  | N                               |                                  |                                                    |
+| ``asfreq``                  | P                               | ``how``, ``normalize``,          | See ``resample``                                   |
```
Comment (sfc-gh-azhan) [inline docs/source/modin/supported/series_supported.rst:103]: Can we test if those arguments are used, we raise NotImplementedError?

Comment (sfc-gh-nkrishna) [reply docs/source/modin/supported/series_supported.rst:103]: Good point, added a block to raise NotImplementedError for unused arguments. If an aggregation that is not supported is passed in (like `bfill`), the resample code itself will ensure that NotImplementedError is raised.

**Review by sfc-gh-azhan — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/docstrings/base.py [hunk 6]
```diff
@@ -462,6 +462,49 @@ def apply():
...
+            (note this does not fill NaNs that already were present).
+
+        Returns
+        -------
+        Snowpark pandas :class:`~snowflake.snowpark.modin.pandas.DataFrame` or Snowpark pandas :class:`~snowflake.snowpark.modin.pandas.Series`
+
+        Notes
+        -----
+        This implementation calls `resample` with the `first` aggregation. `asfreq`
+        is only support on DataFrame/Series with DatetimeIndex, and only
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/docstrings/base.py:506]: Nit:
```suggestion
        is only supported on DataFrame/Series with DatetimeIndex, and only
```

File: src/snowflake/snowpark/modin/plugin/docstrings/base.py [hunk 7]
```diff
@@ -462,6 +462,49 @@ def apply():
...
+        the `freq` and `method` parameters are currently supported.
         """
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/docstrings/base.py:508]: Can we also add the examples section?

File: src/snowflake/snowpark/modin/pandas/base.py [hunk 8]
```diff
@@ -1045,21 +1045,25 @@ def error_raiser(msg, exception):
...
+        freq: str,
+        method: FillnaOptions | None = None,
+        how: str | None = None,
+        normalize: bool = False,
+        fill_value: Scalar = None,
     ):  # noqa: PR01, RT01, D200
         """
         Convert time series to specified frequency.
         """
-        # TODO: SNOW-1119855: Modin upgrade - modin.pandas.base.BasePandasDataset
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/pandas/base.py:1055]: SNOW-1119855 is not resolved yet. Are we sure we want to remove this TODO?

File: src/snowflake/snowpark/modin/plugin/docstrings/base.py (hunk 6 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/docstrings/base.py:506]: Fixed in both frontend + QC docstrings

File: src/snowflake/snowpark/modin/pandas/base.py (hunk 8 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/pandas/base.py:1055]: This TODO/ticket is in several places, but the frontend method now matches Modin so we do want to remove the TODO here.

File: src/snowflake/snowpark/modin/plugin/docstrings/base.py (hunk 7 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/docstrings/base.py:538]: Good catch, done!

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks Naren!
